### ISSUE TITLE
Missing final part

### ISSUE THREAD

**Issue by alex-statsig** [relay-tools/fetch-multipart-graphql#48]
I have the following parts of a multipart response:

```
[13:44:34.035] INFO: content-type: application/json; charset=utf-8

{"hasNext":true,"incremental":[{"path":["node","filtered_gates","nodes",13],"label":"ConfigTableListItem_config$defer$TagSelector_tags","data":{"__isIHasTags":"Gate","tags":[]}}]}
---

[13:44:34.035] INFO: content-type: application/json; charset=utf-8

{"hasNext":true,"incremental":[{"path":["node","filtered_gates","nodes",14],"label":"ConfigTableListItem_config$defer$TagSelector_tags","data":{"__isIHasTags":"Gate","tags":[]}}]}
---

[13:44:34.038] INFO: content-type: application/json; charset=utf-8

{"hasNext":false,"incremental":[{"path":["node","filtered_gates","nodes",10],"label":"ConfigTableListItem_config$defer$TagSelector_tags","data":{"__isIHasTags":"Gate","tags":[{"id":"2Qw62B1sA8UESsE9efZBrf","name":"console"}]}},{"path":["node","filtered_gates","nodes",11],"label":"ConfigTableListItem_config$defer$TagSelector_tags","data":{"__isIHasTags":"Gate","tags":[{"id":"4qykikJxPsJaoRm2BgPr7D","name":"experiments"},{"id":"UTiAIZo5n60loea5lFkkG","name":"Unreleased"}]}}]}
-----
```

The final part is not being included in the response (adding some debug logging, it seems PatchResolver's onResponse is being called with an empty array for the final part). I suspect it may have to do with the delimeter being "-----" for the final part? This is coming from @apollo/server v4, I'm trying to implement this into my relay network layer.

**Comment by alex-statsig** [comment]
Upon further reading of the spec (https://www.w3.org/Protocols/rfc1341/7_2_Multipart.html), it seems correct that the final boundary is followed by "--(boundary text)--" (where boundary text is "-" in this case). However this library only seems to use the final boundary if I conclude with "---" instead

**Comment by maraisr** [comment]
If it helps at all, here is an example getting defer/stream working in relay ~> https://github.com/maraisr/meros/blob/main/examples/relay-with-helix/app/relay.ts

It might be a tad out dated with new spec, so you may need to massage the response. For example, "hasNext" relay has no idea, need to put that in `extensions.is_final`.

**Comment by alex-statsig** [comment]
> If it helps at all, here is an example getting defer/stream working in relay ~> https://github.com/maraisr/meros/blob/main/examples/relay-with-helix/app/relay.ts
> 
> Relay also doesn't implement the spec, it's sort of the old one. So you'd need to massage the response. For example, "hasNext" relay has no idea, you need to put that in `extensions.is_final`. "incremental" needs to be unwrapped and put into a `path`, and `data` properties. (No need to change your server, just massage in your fetch function)

Yeah I noticed this mismatch as well, thanks for confirming that. I'll have to take a closer look at your example, landed on something similar:
```
execute: (req, vars, cache) => {
          return Observable.create((sink) => {
            fetchMultipart(
              '/api/graphql',
              {
                method: 'POST',
                headers: {
                  'content-type': 'application/json',
                  Accept:
                    'application/json; charset-utf-8, multipart/mixed; deferSpec=20220824',
                },
                body: JSON.stringify({
                  query: req.text,
                  variables: vars,
                }),
                credentials: 'same-origin',
                onNext: (res) => {
                  const parts = res as (
                    | {
                        hasNext?: boolean;
                        incremental: {
                          path: string[];
                          label: string;
                          data: PayloadData;
                        }[];
                      }
                    | { data: PayloadData; hasNext?: boolean }
                  )[];
                  const formatted = parts.flatMap((part) => {
                    if ('incremental' in part) {
                      return part.incremental;
                    }
                    return part;
                  });
                  sink.next(formatted);
                },
                onError: (err) => sink.error(err as Error),
                onComplete: (arg) => {
                  sink.complete();
                },
              },
            );
          });
        },
```

Did you notice a need to set extensions.is_final based on that hasNext property? In my implementation I ignored the "hasNext" field, but perhaps that wouldn't work for @stream, only @defer? I hadn't noticed any issues yet (certainly got my main query to load while a subfragment suspended until the later parts loaded in).

> 
> And in terms of empty responses, not sure where the bug is, but with relay you'll need to `subscriber.next(null)`.

Hmm why do you need this?
I explicitly logged the parts passed to "onNext" handler given to fetchMultipart and did not see the last part (nothing to do with passing to relay, the final "part" just isn't making it to my handler). I was able to "hack" around this by trimming the terminating multipart sequence "-----" from my backend to "---", and now the final part is loaded fine. Certainly seems a bug in how "fetchMultipart" parses the terminating multipart sequence.

**Comment by maraisr** [comment]
out of curiosity does it work if you use use [`meros`](https://github.com/maraisr/meros)?

**Comment by robrichard** [comment]
@alex-statsig thanks for opening this issue. I believe this should be fixed now in v3.1.1. Please let me know if this works for you.

**Comment by alex-statsig** [comment]
> @alex-statsig thanks for opening this issue. I believe this should be fixed now in v3.1.1. Please let me know if this works for you.

It does seem fixed now, thanks so much for the quick turnaround.

> out of curiosity does it work if you use use [`meros`](https://github.com/maraisr/meros)?

Sorry I didn't get a chance to try this out, but seems like it should handle it
